flax/linen/recurrent.py

```diff
@@ -497,3 +500,189 @@ def initialize_carry(rng, batch_dims, size, init_fn=zeros):
+    axis, internally the input will be reshaped to `(t, b0...m, d0...dn)`. The feature
+    dimensions should be last (after the time axis is to the front) and are specified by
+    `num_feature_dims`. The batch dimensions will simply be the remaining dimensions that
+    are not the time axis or the feature dimensions.
+
+    **Output shape**
+
+    If `return_sequences` is True, the output will have the same as the input, else
+    the time dimension will be dropped when the last time step is selected.
+    If `return_state` is True, the output will be a tuple of the output and the
+    carry.
+    """
+
+    if num_feature_dims < 1:
+      raise ValueError(f'num_feature_dims must be >= 1, got: {num_feature_dims}')
+    elif num_feature_dims >= len(x.shape):
+      raise ValueError(
+        f'\'num_feature_dims\' dimensions must be smaller than ' 
+        f'number of dimensions, got num_feature_dims \'{num_feature_dims}\' '
+        f'for input of shape \'{x.shape}\''
+      )
+
+    num_batch_dims = x.ndim - (num_feature_dims + 1) # features dims + time dim
+
+    # move time dimension to the first axis
+    x = x.swapaxes(0, time_axis)
+
+    # WARNING: this definition of initialization state
+    # is known to be problematic: https://github.com/google/flax/issues/652#issuecomment-1124216543
+    # Epecifically, this module will not update its cache if 
+    # 'params' are mutable, which can arise if users use `mutable=True`.
+    is_initializing = self.is_mutable_collection("params")
+
+    # get carry
+    if initial_state is not None:
+      initial_carry = initial_state
+    elif stateful and self.has_variable('cache', 'carry'):
+      initial_carry = self.get_variable('cache', 'carry')
+    else:
+      if num_feature_dims == 1:
+        size = x.shape[-1] # features dim
+      else:
+        size = x.shape[-num_feature_dims:] # spatial + feature dim
+      
+      initial_carry = cell.initialize_carry(
+        init_key,
+        batch_dims=x.shape[1 : 1 + num_batch_dims], 
+        size=size,
+      )
+
+    scan = transforms.scan(
+      cell.__class__.__call__,
+      in_axes=0, out_axes=0,
+      reverse=go_backwards,
+      # scan args
+      variable_axes=variable_axes,
+      variable_broadcast=variable_broadcast,
+      variable_carry=variable_carry,
+      split_rngs=split_rngs,
+    )
+
+    carry, x = scan(cell, initial_carry, x)
+
+    # if the module is initializing keep the initial carry
+    if is_initializing:
+      carry = initial_carry
+
+    if stateful:
+      self.put_variable('cache', 'carry', carry)
+
+    # revert time axis to its original position
+    x = x.swapaxes(0, time_axis)
+
+    # select last time step
+    if not return_sequences:
+      # build slices
+      slices: List[Union[slice, int]] = [slice(None)] * x.ndim
+      slices[time_axis] = -1
+      # apply slices
+      x = x[tuple(slices)]
+
+    if return_state:
+      return x, carry
+    else:
+      return x
+
+class RNN(RNNBase):
```

Comment:
What does the BaseRNN vs RNN class distinction do exactly?

Reply:
`RNNBase` is a base class exposes the `rnn_forward` method that contains the main logic that all rnn implementation can use to define their `__call__`. The `RNN` class is a very generic instance that accepts an arbitrary `RNNCellBase`.